### ISSUE TITLE
Rename "Host Name" field to "Site Key"

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#18813]
We need to make editing a hostName a much more extreme action in dotCMS.  

- By default, a hostName should not look editable, a user should have to click an edit button that pops up a warning saying "Editing your site key can have unintended consequences"

- we should disallow non-dns characters, including spaces, in hostNames

- If you take a look at the Persona > Keytag field, you will see a similar control to the one we envision for this.

**Comment by victoralfaro-dotcms** [comment]
First time:
```7:10:00.843  INFO  util.PushPublishLogger - class com.dotcms.publishing.PublisherAPIImpl : [BundleID: 01FEPW7YNR4CE69YW21RM7DS50] Started Publishing Task
17:10:58.910  WARN  pool.HikariPool - jdbc/dotCMSPool - Thread starvation or clock leap detected (housekeeper delta=1m27s148ms).
17:10:58.910  WARN  pool.HikariPool - HikariPool-1 - Thread starvation or clock leap detected (housekeeper delta=1m8s84ms).
17:10:58.910  WARN  pool.HikariPool - HikariPool-2 - Thread starvation or clock leap detected (housekeeper delta=1m8s89ms).
17:11:05.647  INFO  publishing.PublisherAPIImpl - Writing bundle.xml file
17:11:20.552  INFO  publishing.PublisherAPIImpl - Start of Bundler: DependencyBundler
17:11:21.397  INFO  caffine.CaffineCache - ***	 Building Cache : pushedassetscache, size:5000,Concurrency:32
17:11:24.048  INFO  dependencies.ConcurrentDependencyProcessor - DependencyProcessor Finished
17:11:26.421  INFO  publishing.PublisherAPIImpl - End of Bundler: DependencyBundler
17:11:28.246  INFO  publishing.PublisherAPIImpl - Start of Bundler: HostBundler
17:11:28.897  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.bundler.HostBundler : Status Update: Bundling.
17:11:28.969  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.bundler.HostBundler : Status Update: Bundling.
17:11:31.085  INFO  publishing.PublisherAPIImpl - End of Bundler: HostBundler
17:11:31.086  INFO  publishing.PublisherAPIImpl - Start of Bundler: ContentBundler
17:11:31.119  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.bundler.ContentBundler : Status Update: Bundling.
17:11:31.246  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.bundler.ContentBundler : Status Update: Bundling.
17:11:31.296  INFO  publishing.PublisherAPIImpl - End of Bundler: ContentBundler
17:11:31.296  INFO  publishing.PublisherAPIImpl - Start of Bundler: FolderBundler
17:11:31.297  INFO  publishing.PublisherAPIImpl - End of Bundler: FolderBundler
17:11:31.298  INFO  publishing.PublisherAPIImpl - Start of Bundler: TemplateBundler
17:11:31.298  INFO  publishing.PublisherAPIImpl - End of Bundler: TemplateBundler
17:11:31.299  INFO  publishing.PublisherAPIImpl - Start of Bundler: ContainerBundler
17:11:31.299  INFO  publishing.PublisherAPIImpl - End of Bundler: ContainerBundler
17:11:31.299  INFO  publishing.PublisherAPIImpl - Start of Bundler: RuleBundler
17:11:31.300  INFO  publishing.PublisherAPIImpl - End of Bundler: RuleBundler
17:11:31.300  INFO  publishing.PublisherAPIImpl - Start of Bundler: LinkBundler
17:11:31.301  INFO  publishing.PublisherAPIImpl - End of Bundler: LinkBundler
17:11:31.301  INFO  publishing.PublisherAPIImpl - Start of Bundler: ContentTypeBundler
17:11:31.302  INFO  publishing.PublisherAPIImpl - End of Bundler: ContentTypeBundler
17:11:31.302  INFO  publishing.PublisherAPIImpl - Start of Bundler: RelationshipBundler
17:11:31.303  INFO  publishing.PublisherAPIImpl - End of Bundler: RelationshipBundler
17:11:31.303  INFO  publishing.PublisherAPIImpl - Start of Bundler: LanguageVariablesBundler
17:11:31.324  INFO  publishing.PublisherAPIImpl - End of Bundler: LanguageVariablesBundler
17:11:31.325  INFO  publishing.PublisherAPIImpl - Start of Bundler: WorkflowBundler
17:11:31.325  INFO  publishing.PublisherAPIImpl - End of Bundler: WorkflowBundler
17:11:31.325  INFO  publishing.PublisherAPIImpl - Start of Bundler: LanguageBundler
17:11:31.373  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.bundler.LanguageBundler : Status Update: Bundling.
17:11:31.420  INFO  publishing.PublisherAPIImpl - End of Bundler: LanguageBundler
17:11:31.421  INFO  publishing.PublisherAPIImpl - Start of Bundler: BundleXMLAsc
17:11:31.421  INFO  publishing.PublisherAPIImpl - End of Bundler: BundleXMLAsc
17:11:31.422  INFO  publishing.PublisherAPIImpl - Start of Bundler: CategoryBundler
17:11:31.422  INFO  publishing.PublisherAPIImpl - End of Bundler: CategoryBundler
```

**Comment by victoralfaro-dotcms** [comment]
Next attempt to PP
```
17:11:42.925  INFO  util.PushPublishLogger - class com.dotcms.publisher.business.PublisherQueueJob : Status update: Failed to update bundle audit status.
17:11:42.968  INFO  business.PublisherQueueJob - ===========================================================
17:11:42.968  INFO  business.PublisherQueueJob - For bundle '01FEPW7YNR4CE69YW21RM7DS50':
17:11:42.968  INFO  business.PublisherQueueJob - -> Status             : FAILED_TO_SEND_TO_ALL_GROUPS [4]
17:11:42.968  INFO  business.PublisherQueueJob - -> Re-publish attempts: 1 out of 3
17:11:42.968  INFO  business.PublisherQueueJob - ===========================================================
17:11:42.968  INFO  business.PublisherQueueJob -
17:11:42.968  INFO  business.PublisherQueueJob - Found 1 bundle(s) to process.
17:11:42.968  INFO  business.PublisherQueueJob -
17:11:42.968  INFO  business.PublisherQueueJob - Processing bundle: ID: 01FEPW7YNR4CE69YW21RM7DS50. Status: 4. Publish Date: 2021-09-03 17:08:00.0
17:11:42.968  INFO  util.PushPublishLogger - class com.dotcms.publisher.business.PublishAuditStatus : Status Update: Pending.
17:11:42.968  INFO  util.PushPublishLogger - class com.dotcms.publisher.business.PublisherQueueJob : Pre-publish work started.
[class com.dotcms.publisher.pusher.PushPublisher]
17:11:42.970  INFO  util.PushPublishLogger - class com.dotcms.publisher.business.PublisherQueueJob : Pre-publish work complete.
17:11:42.970  INFO  util.PushPublishLogger - class com.dotcms.publishing.PublisherAPIImpl : [BundleID: 01FEPW7YNR4CE69YW21RM7DS50] Started Publishing Task
17:11:56.295  INFO  publishing.PublisherAPIImpl - Retrying bundle: 01FEPW7YNR4CE69YW21RM7DS50, we don't need to run bundlers again
```

**Comment by bryanboza** [comment]
Here some problems after those changes: 

- We need to fire a web-socket event to update the site selector component after rename the host. 

- The site must be unique, since we are allowing to rename the site, we need to validate if the new site-name already exist before allow to rename, we are doing it locally, but If you rename the site and then push to a remote server. We are not validating it and this will cause problems, this because we have a way to have two sites with the same name
![image](https://user-images.githubusercontent.com/2641437/144461372-803e877e-c656-4fd6-8772-b552e7c9fa19.png)

**Comment by fabrizzio-dotCMS** [comment]
@bryanboza 
Regarding the input validation. There is a property in the backend that validates all that. 

```
private void validateSite(Contentlet contentlet) {
        if (Config.getBooleanProperty("site.key.dns.validation", false)) {
            final String siteKey = (String) contentlet.get(Host.HOST_NAME_KEY);
            if (!UtilMethods.isSet(siteKey) || !dnsPattern.matcher(siteKey).find()) {
                throw new DotContentletValidationException(
                        String.format("Site key %s doesn't match a valid dns format.", siteKey));
            }
        }
    }
```
if you add this to your properties 

site.key.dns.validation=true

You'll see it. 
But it was agreed during our review meeting that the validation must be turned off by default to avoid affecting customers that could have site names that break the pattern.

**Comment by nollymar** [comment]
[This](https://github.com/dotCMS/core/issues/18813#issuecomment-984857341) probably will need doc @rweiner

**Comment by fabrizzio-dotCMS** [comment]
> Here some problems after those changes:
> 
> * We need to fire a web-socket event to update the site selector component after rename the host.
> * The site must be unique, since we are allowing to rename the site, we need to validate if the new site-name already exist before allow to rename, we are doing it locally, but If you rename the site and then push to a remote server. We are not validating it and this will cause problems, this because we have a way to have two sites with the same name
>   ![image](https://user-images.githubusercontent.com/2641437/144461372-803e877e-c656-4fd6-8772-b552e7c9fa19.png)
> * Also, we are not validating the blank spaces, we should not allow to set site names with blank spaces and neither special characters. Right now I'm able to set this name
>   ![image](https://user-images.githubusercontent.com/2641437/144464786-0f558bc4-333b-45f1-9b9c-2564e6971fdd.png)

@bryanboza 
I was able to reproduce this issue on 21-11 (and it wasn't easy to do that)
It took me many attempts creating sites then pushing them then deleting them on the sender and creating a new one (and pushing it) using a  name already used in the past but no longer present on the sender. 

I think what I did was 
Create a site pushing it. then creating a second site and pushing it again. Once both sites exist on the receiver I delete one of them on the sender and rename the other to have whatever name I had on the first and then push it. 

It takes many attempts to see this surface. 

![Screen Shot 2021-12-03 at 11 57 30](https://user-images.githubusercontent.com/37148755/144650924-48db369f-4871-49c9-b42f-330ba5239497.png)


It feels like an indexing thing to me.

**Comment by nollymar** [comment]
@bryanboza  if it's happening in older versions, a new card should be created to fix it because these changes didn't break that

**Comment by bryanboza** [comment]
@nollymar and @fabrizzio-dotCMS ok seems we are able to reproduce in older versions, but as fabricio said, It is not easy to reproduce, after allow to rename now it is really easy to reproduce and this can cause a lot of problems to the environment, then not sure if the case maybe require some priority or not for now. @wezell 

Also, we have two more points pending on this card.

**Comment by fabrizzio-dotCMS** [comment]
> @nollymar and @fabrizzio-dotCMS ok seems we are able to reproduce in older versions, but as Fabricio said, It is not easy to reproduce and those are not the same steps, after allow to rename now it is really easy to reproduce, we have the steps how to reproduce, and this can cause a lot of problems to the environment and of course to support, then not sure if the case maybe require some priority or not for now. @wezell
> 
> Also, we have two more points pending on this card.

One of thise points I already clarified in a comment up here. 

I think the only visible issue is the Websock

**Comment by bryanboza** [comment]
@fabrizzio-dotCMS here the easy way to reproduce the issue: 

- Start and configure your push env (both nodes)
- On sender create two new sites "cody" and "rex"
- Push both 
- Now delete "cody" 
- Rename "rex" to "cody"
- Push "cody" again

**Comment by fabrizzio-dotCMS** [comment]
> @fabrizzio-dotCMS here the easy way to reproduce the issue:
> 
> * Start and configure your push env (both nodes)
> * On sender create two new sites "cody" and "rex"
> * Push both
> * Now delete "cody"
> * Rename "rex" to "cody"
> * Push "cody" again
> * Check the receiver side
@bryanboza 
true it was quite simple to reproduce however it's an old issue.  I was able to do it on `21-11`

**Comment by jcastro-dotcms** [comment]
Excluding from LTS releases given that back-porting the required code changes can potentially break the system.

**Comment by nollymar** [comment]
**Note for Doc:** In our last meeting, we agreed to turn on the property `site.key.dns.validation` by default. 
PR: #21387

**Comment by fabrizzio-dotCMS** [comment]
Note for QA:
The site selector was broken because of a missing boolean field (with null value) that by default should be interpreted as 'false'
I added a small fix that basically ensures that for all fields with a dataType of type (BOOLEAN,FLOAT,INTEGER) when No value is set. The field gets created with a default value. 
for boolen data-types the field gets initiallized with 'false'
for integer data-type the field gets initiallized with 0
for float point data-points the field gets initiallized with 0.0

What needs to be tested here? 

First make sure the Site seletor works as expected.

Here's the basic scenario. 
You create a Content-Type add a Field that supports any of the data-types mentioned above e.g. 
Then create an instance and hit save (adding no explicit value). Make sure the value gets created in the json field with the expected default value.

You could use a query (via Query tool) to corroborate the instances of contet match the expected default value.

**Comment by alfredo-dotcms** [comment]
when trying to create a Site, the dialog freeze and no site is created
![image](https://user-images.githubusercontent.com/37185433/145851695-b2228a87-08ae-41a4-a918-342b9e7f9e60.png)

**Comment by alfredo-dotcms** [comment]
![image](https://user-images.githubusercontent.com/37185433/145854347-f31be79d-8ec1-4174-9854-cb861b8f0fdc.png)

looks good

**Comment by bryanboza** [comment]
Fixed, tested on release-21.12 // Docker // FF. New card related to site selector here #21418

**Comment by bryanboza** [comment]
We need some extra work here: 
- If you try to use a non-dns characters, including spaces in hostNames we are printing a multiple lines of error, we need to catch this exception and leave just a line error 
https://gist.github.com/bryanboza/f10afa571077098c72a416e751e12eb4

- A new case related to site selector here #21418

**Comment by nollymar** [comment]
@bryanboza during a call with @wezell we agreed that the issues you mentioned aren't blockers. So, the first item won't be fixed as it is the way we handle contentlet exceptions. About the second error, there is already a PR for next release.

**Comment by bryanboza** [comment]
According to the last comment, then we are ready with this card. Fixed, tested on release-21.12 // Docker // FF
